domain/network/state/space.go

```diff
@@ -42,85 +41,92 @@ func (st *State) AddSpace(
-INSERT INTO space (uuid, name)
-VALUES (?, ?)`
+	insertSpaceStmt, err := sqlair.Prepare(
+		`INSERT INTO space (uuid, name) VALUES ($Space.uuid, $Space.name)`, Space{},
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	insertProviderStmt := `
-INSERT INTO provider_space (provider_id, space_uuid)
-VALUES (?, ?)`
+	insertProviderStmt, err := sqlair.Prepare(
+		`INSERT INTO provider_space (provider_id, space_uuid) VALUES ($ProviderSpace.provider_id, $ProviderSpace.space_uuid)`,
+		ProviderSpace{},
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	subnetBinds, subnetVals := database.SliceToPlaceholder(subnetIDs)
-	findFanSubnetsStmt := fmt.Sprintf(`
-SELECT subject_subnet_uuid
+	findFanSubnetsStmt, err := sqlair.Prepare(`
+SELECT subject_subnet_uuid AS &Subnet.uuid
 FROM   subnet_association
-WHERE  association_type_id = 0 AND associated_subnet_uuid IN (%s)`, subnetBinds)
+WHERE  association_type_id = 0 AND associated_subnet_uuid IN ($S[:])`, sqlair.S{}, Subnet{})
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	checkInputSubnetsStmt := fmt.Sprintf(`
-SELECT uuid
+	checkInputSubnetsStmt, err := sqlair.Prepare(`
+SELECT &Subnet.uuid
 FROM   subnet
 JOIN   subnet_type
 ON     subnet.subnet_type_id = subnet_type.id
-WHERE  subnet_type.is_space_settable = FALSE AND subnet.uuid IN (%s)`, subnetBinds)
+WHERE  subnet_type.is_space_settable = FALSE AND subnet.uuid IN ($S[:])`, sqlair.S{}, Subnet{})
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
 		// We must first check on the provided subnet ids to validate
 		// that are of a type on which the space can be set.
-		nonSettableSubnets, err := tx.QueryContext(ctx, checkInputSubnetsStmt, subnetVals...)
+
+		subnetIDsInS := sqlair.S{}
+		for _, sid := range subnetIDs {
+			subnetIDsInS = append(subnetIDsInS, sid)
+		}
+
+		var nonSettableSubnets []Subnet
+		err := tx.Query(ctx, checkInputSubnetsStmt, subnetIDsInS).GetAll(&nonSettableSubnets)
 		if err != nil {
 			return errors.Annotatef(err, "checking if there are fan subnets for space %q", uuid)
 		}
-		defer func() { _ = nonSettableSubnets.Close() }()
 		// If any row is returned we must fail with the returned fan
 		// subnet uuids.
-		uniqueErrorSubnetUUIDs := make(map[string]string)
-		var nonSettableUUIDs []string
-		for nonSettableSubnets.Next() {
-			var fanSubnetID string
-			err = nonSettableSubnets.Scan(&fanSubnetID)
-			if err != nil {
-				return errors.Trace(err)
+		if len(nonSettableSubnets) > 0 {
+			uniqueErrorSubnetUUIDs := make(map[string]string)
+			var nonSettableUUIDs []string
+			for _, nonSettableSubnet := range nonSettableSubnets {
+				fanSubnetID := nonSettableSubnet.UUID
+				if _, ok := uniqueErrorSubnetUUIDs[fanSubnetID]; !ok {
+					uniqueErrorSubnetUUIDs[fanSubnetID] = fanSubnetID
+					nonSettableUUIDs = append(nonSettableUUIDs, fanSubnetID)
+				}
 			}
```

Comment:
We should extract this logic into another function that allows us to test it in the future (in addition to: reduces our cyclomatic complexity)

Reply:
Moved into a separate function `uniqueSubnetUUIDs`. This logic is also repeated a few lines down so it has the added benefit of reuse. I've not exported the function so I've not added any new tests for it.